### PR TITLE
Impl Add<Month> + Sub<Month> for NaiveDateTime

### PR DESCRIPTION
**Pull request by avantgardnerio**
### Thanks for contributing to chrono!

- [x] Have you added yourself and the change to the [changelog]? (Don't worry
      about adding the PR number)
- [ ] If this pull request fixes a bug, does it add a test that verifies that
      we can't reintroduce it?

[changelog]: ../CHANGELOG.md

### CODE REVIEW

**Comment by djc** [comment]
This seems reasonable. I think we should probably also add `checked_add_months()` and `checked_sub_months()`, and I'd like to express the new `Add` and `Sub` implementation in terms of `NaiveDate::checked_add_months()` and `checked_sub_months()` as well, to reduce the amount of layering and make the panicking behavior more explicit.

**Comment by avantgardnerio** [comment]
@djc I added these functions, and I think they meet the requirements listed above. I inadvertently found an edge-case that was panicking, and added a test and fixed that too (see build logs for details).

File: src/naive/date.rs [hunk 1]
```diff
@@ -629,22 +629,28 @@ impl NaiveDate {
...
-    ///
     /// ```
     /// # use chrono::{NaiveDate, Months};
     /// assert_eq!(
     ///     NaiveDate::from_ymd(2022, 2, 20).checked_sub_months(Months::new(6)),
     ///     Some(NaiveDate::from_ymd(2021, 8, 20))
     /// );
     /// ```
+    ///
+    /// Returns `None` if the resulting date would be out of range.
```
Comment (djc) [inline src/naive/date.rs:640]: I'd prefer to keep this explanation before the examples (and I think that's common style for rustdoc).

File: src/naive/datetime/mod.rs [hunk 2]
```diff
@@ -533,6 +533,34 @@ impl NaiveDateTime {
...
+    ///
+    /// ```
+    /// use std::str::FromStr;
+    /// use chrono::{Months, NaiveDateTime};
+    /// let dt = |s| NaiveDateTime::from_str(s).unwrap();
+    /// let m = Months::new(core::i32::MAX as u32 + 1);
+    /// assert_eq!(dt("2014-01-01T01:00:00").checked_add_months(m), None);
+    /// ```
+    pub fn checked_add_months(self, rhs: Months) -> Option<NaiveDateTime> {
+        self.date().checked_add_months(rhs).map(|dt| dt.and_time(self.time()))
```
Comment (djc) [inline src/naive/datetime/mod.rs:561]: I think this can be simplified to

```rust
self.date = self.date.checked_add_months(rhs)?;
```

I definitely think we should not be using the `date()` and `time()` accessors here.

File: src/naive/datetime/mod.rs [hunk 3]
```diff
@@ -1289,6 +1345,37 @@ impl AddAssign<OldDuration> for NaiveDateTime {
...
+    /// assert_eq!(dt("2014-01-01T01:00:00").unwrap() + m(1), dt("2014-02-01T01:00:00").unwrap());
+    /// assert_eq!(dt("2014-01-01T00:02:00").unwrap() + m(11), dt("2014-12-01T00:02:00").unwrap());
+    /// assert_eq!(dt("2014-01-01T00:00:03").unwrap() + m(12), dt("2015-01-01T00:00:03").unwrap());
+    /// assert_eq!(dt("2014-01-01T00:00:04").unwrap() + m(13), dt("2015-02-01T00:00:04").unwrap());
+    /// assert_eq!(dt("2014-01-31T00:05:00").unwrap() + m(1), dt("2014-02-28T00:05:00").unwrap());
+    /// assert_eq!(dt("2020-01-31T06:00:00").unwrap() + m(1), dt("2020-02-29T06:00:00").unwrap());
+    /// ```
+    fn add(self, rhs: Months) -> Self::Output {
+        let new_date = self.date() + rhs;
+        new_date.and_time(self.time())
```
Comment (djc) [inline src/naive/datetime/mod.rs:1375]:
```rust
Self {
    date: self.date.checked_add_months(rhs).unwrap(),
    time: self.time,
}
```

File: src/naive/date.rs [hunk 4]
```diff
@@ -629,22 +629,28 @@ impl NaiveDate {
...
+    /// ```
     pub fn checked_sub_months(self, months: Months) -> Option<Self> {
         if months.0 == 0 {
             return Some(self);
         }
 
-        // Copy `i32::MIN` here so we don't have to do a complicated cast
-        match months.0 <= 2_147_483_648 {
+        // Copy `i32::MAX` here so we don't have to do a complicated cast
+        match months.0 <= 2_147_483_647 {
```
Comment (djc) [inline src/naive/date.rs:653]: Ah, missed that we're only making this negative after the cast to `i32`.

File: src/naive/datetime/mod.rs (hunk 3 above)
Comment (djc) [reply src/naive/datetime/mod.rs:1375]: This doesn't seem to be resolved?

Comment (avantgardnerio) [reply src/naive/datetime/mod.rs]: Sorry, I fixed it in the other two places and forgot to fix it here. It is resolved now.

**Review by djc — APPROVED** [review]

File: src/naive/datetime/mod.rs [hunk 5]
```diff
@@ -606,6 +634,34 @@ impl NaiveDateTime {
         Some(NaiveDateTime { date, time })
     }
 
+    /// Subtracts given `Months` to the current date and time.
```
Comment (djc) [inline src/naive/datetime/mod.rs:637]: Sorry, minor nit I forgot to mention: "Subtracts .. to" should use have "from" instead of "to".

**Review by djc — APPROVED** [review]

File: src/naive/datetime/mod.rs [hunk 6]
```diff
@@ -533,6 +533,34 @@ impl NaiveDateTime {
...
+    ///
+    /// # Example
+    ///
+    /// ```
+    /// use std::str::FromStr;
+    /// use chrono::{Months, NaiveDateTime};
+    /// let dt = |s| NaiveDateTime::from_str(s).unwrap();
+    /// let ca = |dt: NaiveDateTime, m: Months| dt.checked_add_months(m).unwrap();
+    /// let m = Months::new;
+    /// assert_eq!(ca(dt("2014-01-01T01:00:00"), m(1)), dt("2014-02-01T01:00:00"));
```
Comment (esheppa) [inline src/naive/datetime/mod.rs:548]: I'm a little uncomfortable with this style of code (albeit similar to what is used in the private tests) being in the public docs. IMO this would look better if we just used the full functions directly, with more intermediate variables if needed for readability. 

I've set up https://github.com/avantgardnerio/chrono/pull/1 as one possibility here, but I'm not set on that option, and keen to have a discussion on the merits of each approach.

**Review by esheppa — CHANGES_REQUESTED** [review]
Thanks @avantgardnerio - good idea to add this to `NaiveDateTime` (I'm keen to set up a PR to add it to `DateTime<Tz>` shortly after this :smile:) - I've set up a PR to your branch with a proposal for a change to the documentation, but apart from that this looks great!

File: src/naive/datetime/mod.rs (hunk 6 above)
Comment (djc) [reply src/naive/datetime/mod.rs:548]: I agree that this makes the examples harder to understand (and somewhat unidiomatic).

Comment (avantgardnerio) [reply src/naive/datetime/mod.rs]: I think they are all fixed now. Please ping me if I missed any!

**Comment by avantgardnerio** [comment]
Thanks all! :tada:
